util/cstruct/cstruct.go

```diff
@@ -0,0 +1,208 @@
+	d.off += size
+	return nil
+}
+
+// Err returns the first error that was encountered by this Decoder.
+func (d *Decoder) Err() error {
+	return d.err
+}
+
+// Offset returns the current read offset for data in the buffer.
+func (d *Decoder) Offset() int {
+	return d.off
+}
+
+// Byte reutrns a single byte from the buffer.
+func (d *Decoder) Byte() byte {
+	if d.err != nil {
+		return 0
+	}
+
+	var b [1]byte
+	if err := d.getField(b[:]); err != nil {
+		d.err = err
+		return 0
+	}
+	return b[0]
+}
+
+// Byte reutrns a number of bytes from the buffer based on the size of the
+// input slice. No padding is applied.
+//
+// If an error is encountered or this Decoder has previously encountered an
+// error, no changes are made to the provided buffer.
+func (d *Decoder) Bytes(b []byte) {
+	if d.err != nil {
+		return
+	}
+
+	// No padding for byte slices
+	size := len(b)
+	if d.off+size >= len(d.b) {
+		d.err = io.EOF
+		return
+	}
+	copy(b, d.b[d.off:d.off+size])
+	d.off += size
+}
+
+// Uint16 returns a uint16 decoded from the buffer.
+func (d *Decoder) Uint16() uint16 {
+	if d.err != nil {
+		return 0
+	}
+
+	var buf [2]byte
+	if err := d.getField(buf[:]); err != nil {
+		d.err = err
+		return 0
+	}
+	return endian.Native.Uint16(buf[:])
+}
+
+// Uint32 returns a uint32 decoded from the buffer.
+func (d *Decoder) Uint32() uint32 {
+	if d.err != nil {
+		return 0
+	}
+
+	var buf [4]byte
+	if err := d.getField(buf[:]); err != nil {
+		d.err = err
+		return 0
+	}
+	return endian.Native.Uint32(buf[:])
+}
+
+// Uint64 returns a uint64 decoded from the buffer.
+func (d *Decoder) Uint64() uint64 {
+	if d.err != nil {
+		return 0
+	}
+
+	var buf [8]byte
+	if err := d.getField(buf[:]); err != nil {
+		d.err = err
+		return 0
+	}
+	return endian.Native.Uint64(buf[:])
+}
+
+// Uintptr returns a uintptr decoded from the buffer.
+func (d *Decoder) Uintptr() uintptr {
+	if d.err != nil {
+		return 0
+	}
+
+	switch pointerSize {
+	case 4:
+		return uintptr(d.Uint32())
+	case 8:
+		return uintptr(d.Uint64())
+	default:
+		panic("unsupported")
+	}
+}
+
+// Int16 returns a int16 decoded from the buffer.
+func (d *Decoder) Int16() int16 {
+	if d.err != nil {
+		return 0
+	}
+
+	var buf [2]byte
+	if err := d.getField(buf[:]); err != nil {
+		d.err = err
+		return 0
+	}
+	return int16(endian.Native.Uint16(buf[:]))
+}
+
+// Int32 returns a int32 decoded from the buffer.
+func (d *Decoder) Int32() int32 {
+	if d.err != nil {
+		return 0
+	}
+
+	var buf [4]byte
+	if err := d.getField(buf[:]); err != nil {
+		d.err = err
+		return 0
+	}
+	return int32(endian.Native.Uint32(buf[:]))
+}
+
+// Int64 returns a int64 decoded from the buffer.
+func (d *Decoder) Int64() int64 {
+	if d.err != nil {
+		return 0
+	}
+
+	var buf [8]byte
+	if err := d.getField(buf[:]); err != nil {
+		d.err = err
+		return 0
+	}
+	return int64(endian.Native.Uint64(buf[:]))
+}
+
+// TODO(andrew): remove unsafe with build tags?
+var pointerSize = int(unsafe.Sizeof(uintptr(0)))
```

Comment:
you need neither unsafe nor build tags... e.g. https://github.com/tailscale/tailscale/blob/main/types/pad32/pad32.go#L8-L9

Reply:
This is a neat trick; did something similar to make a `const pointerSize`

Reply:
You're in good company:

```
bradfitz@tsdev:~/hack/go$ git grep -F ">> 63" | grep uint
misc/cgo/errors/testdata/malloc.go:		size = C.size_t(0x7fffffff << (32 * (^uintptr(0) >> 63)))
src/cmd/compile/internal/ssa/poset.go:const uintSize = 32 << (^uint(0) >> 63) // 32 or 64
src/cmd/vendor/golang.org/x/tools/go/analysis/passes/shift/shift.go:		// like ^uint(0) >> 63 for 32/64 bit detection and compatibility.
src/crypto/subtle/constant_time.go:	return int((uint64(uint32(x^y)) - 1) >> 63)
src/fmt/scan.go:	intBits     = 32 << (^uint(0) >> 63)
src/fmt/scan.go:	uintptrBits = 32 << (^uintptr(0) >> 63)
src/image/png/paeth.go:const intSize = 32 << (^uint(0) >> 63)
src/internal/goarch/goarch.go:const PtrSize = 4 << (^uintptr(0) >> 63)
src/math/bits/bits.go:const uintSize = 32 << (^uint(0) >> 63) // 32 or 64
src/math/const.go:	intSize = 32 << (^uint(0) >> 63) // 32 or 64
src/math/fma.go:	sign = uint32(b >> 63)
src/runtime/internal/math/math_test.go:	UintptrSize = 32 << (^uintptr(0) >> 63)
src/runtime/malloc.go:	_64bit = 1 << (^uintptr(0) >> 63) / 2
src/runtime/string_test.go:const intSize = 32 << (^uint(0) >> 63)
src/strconv/atoi.go:const intSize = 32 << (^uint(0) >> 63)
src/strconv/decimal.go:const uintSize = 32 << (^uint(0) >> 63)
src/vendor/golang.org/x/sys/cpu/hwcap_linux.go:	uintSize = int(32 << (^uint(0) >> 63))
```